src/libstore/signature/signer.cc

```diff
@@ -0,0 +1,85 @@
+    {
+        size_t colon = s.find(':');
+        if (colon == std::string::npos || colon == 0)
+            return {"", ""};
+        return {s.substr(0, colon), s.substr(colon + 1)};
+    }
+
+    Signer::Signer() : pubkey(this->getPublicKey()) { }
+    Signer::Signer(PublicKey && pubkey) : pubkey(pubkey) { }
+
+    bool Signer::verifyDetached(const std::string & data, const std::string & sig) {
+        auto ss = split(sig);
+
+        if (std::string(ss.first) != pubkey.key) return false;
+
+        auto sig2 = base64Decode(ss.second);
+        if (sig2.size() != crypto_sign_BYTES)
+            throw Error("signature is not valid");
+
+        return crypto_sign_verify_detached((unsigned char *) sig2.data(),
+            (unsigned char *) data.data(), data.size(),
+            (unsigned char *) pubkey.key.data()) == 0;
+    }
+
+    const PublicKey& Signer::getPublicKey() const {
+        return pubkey;
+    }
+
+    LocalSigner::LocalSigner(SecretKey && privkey) : Signer(privkey.toPublicKey()), privkey(privkey) {
+    }
+
+    std::string LocalSigner::signDetached(std::string_view s) const {
+        return privkey.signDetached(s);
+    }
+
+    RemoteSigner::RemoteSigner(const std::string & remoteServerUrl) : Signer(),
+        serverUrl(remoteServerUrl),
+        _curl_handle(std::unique_ptr<CURL, decltype(&curl_easy_cleanup)>(curl_easy_init(), curl_easy_cleanup)) {
+        if (!_curl_handle) {
+            throw Error("Failed to initialize curl");
+        }
+
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_USERAGENT, "Nix $version");
+        // Signing should be very fast, 5s is already very long.
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_TIMEOUT, 5);
+
+        // The remote URL should not have no trailing /.
+        assert(serverUrl.back() != '/');
+    }
+
+    // Write the HTTP response inside the userdata string buffer provided.
+    static size_t _writeResponse(void* ptr, size_t size, size_t nmemb, void* userdata) {
+        size_t real_size = size *nmemb;
+        std::string* response = static_cast<std::string*>(userdata);
+        response->append(static_cast<char*>(ptr), real_size);
+        return real_size;
+    }
+
+    std::string RemoteSigner::signDetached(std::string_view s) const {
+        std::string detached_signature;
+        std::string hash_to_sign = std::string(s);
+
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_URL, (serverUrl + "/sign").c_str());
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_WRITEFUNCTION, &_writeResponse);
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_WRITEDATA, &detached_signature);
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_POSTFIELDS, hash_to_sign.c_str());
+
+        CURLcode res = curl_easy_perform(_curl_handle.get());
+        if (res != CURLE_OK) {
+            // log the error?
+            throw Error("failed to sign remotely, received non-200 code");
```

Comment:
curl doesn't validate the HTTP response code, this is for stuff like the connection breaking or the server not speaking HTTP.

Reply:
Fixed.